esp-hal/src/dma/mod.rs

```diff
@@ -732,59 +730,67 @@ where
     }
 
     fn available(&mut self) -> usize {
-        if self.last_seen_handled_descriptor_ptr.is_null() {
-            self.last_seen_handled_descriptor_ptr = self.descriptors.as_mut_ptr();
-            return 0;
-        }
-
-        if self.available != 0 {
-            return self.available;
+        let last_in_dscr_ptr = self.rx_impl.last_in_dscr_address() as *mut DmaDescriptor;
+        let last_in_descr = unsafe { last_in_dscr_ptr.read_volatile() };
+
+        if !last_in_descr.is_empty() && last_in_descr.owner() == Owner::Cpu {
+            let len = last_in_descr.len();
+
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
```

Comment:
You can skip this line I think, since you do this at the end of the function anyway